Importing the required libraries

In [1]:
import numpy as np
import pandas as pd

importing the required dataset

In [2]:
Loan_df=pd.read_csv(r"D:\Data Analysis\Projects\Loan Analyser eda\data\india_personal_loan_default_risk_2026.csv")
display(Loan_df)
display(Loan_df.head(5))
display(Loan_df.shape)

,applicant_id,age,gender,marital_status,city_tier,state,employment_type,bank_account_vintage_years,monthly_income_inr,existing_loans_count,...,num_credit_inquiries_last_6m,late_payments_last_12m,loan_amount_requested_inr,loan_tenure_months,loan_purpose,collateral_provided,cibil_score,interest_rate_offered_pct,default_risk_score,default_flag
0,APP106868,22,Male,Single,Tier 3,Haryana,Salaried - PSU,0.5,12000,1,...,2.0,0,149000,48,Personal Expense,No,649,24.84,1.67,0
1,APP124016,35,Female,Divorced,Tier 2,Gujarat,Salaried - Private,0.5,32500,3,...,0.0,1,156000,60,Personal Expense,No,538,25.25,4.47,0
2,APP109668,41,Male,Married,Tier 2,Uttar Pradesh,Salaried - Government,2.8,52500,0,...,0.0,0,70000,36,Wedding,No,636,20.46,0.19,0
3,APP113640,29,Female,Married,Tier 2,Kerala,Salaried - PSU,0.7,15500,1,...,0.0,1,108000,48,Education,No,721,15.27,0.26,0
4,APP114018,31,Male,Married,Tier 1,Punjab,Self-Employed Professional,4.7,269600,2,...,0.0,2,466000,24,Education,No,586,25.52,2.10,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
24995,APP121575,29,Male,Single,Tier 1,Andhra Pradesh,Salaried - Private,0.7,85900,1,...,1.0,1,113000,24,Personal Expense,No,594,25.24,1.00,0
24996,APP105390,49,Female,Single,Tier 1,Punjab,Self-Employed Professional,0.9,142600,1,...,2.0,2,35000,24,Debt Consolidation,No,638,19.78,0.56,0
24997,APP100860,26,Other,Divorced,Tier 1,Haryana,Salaried - Private,3.0,76600,1,...,2.0,0,67000,36,Wedding,No,562,19.11,4.03,0
24998,APP115795,44,Female,Married,Tier 1,Gujarat,Self-Employed Professional,7.3,68200,1,...,2.0,0,296000,84,Medical Emergency,No,665,16.23,0.45,0


,applicant_id,age,gender,marital_status,city_tier,state,employment_type,bank_account_vintage_years,monthly_income_inr,existing_loans_count,...,num_credit_inquiries_last_6m,late_payments_last_12m,loan_amount_requested_inr,loan_tenure_months,loan_purpose,collateral_provided,cibil_score,interest_rate_offered_pct,default_risk_score,default_flag
0,APP106868,22,Male,Single,Tier 3,Haryana,Salaried - PSU,0.5,12000,1,...,2.0,0,149000,48,Personal Expense,No,649,24.84,1.67,0
1,APP124016,35,Female,Divorced,Tier 2,Gujarat,Salaried - Private,0.5,32500,3,...,0.0,1,156000,60,Personal Expense,No,538,25.25,4.47,0
2,APP109668,41,Male,Married,Tier 2,Uttar Pradesh,Salaried - Government,2.8,52500,0,...,0.0,0,70000,36,Wedding,No,636,20.46,0.19,0
3,APP113640,29,Female,Married,Tier 2,Kerala,Salaried - PSU,0.7,15500,1,...,0.0,1,108000,48,Education,No,721,15.27,0.26,0
4,APP114018,31,Male,Married,Tier 1,Punjab,Self-Employed Professional,4.7,269600,2,...,0.0,2,466000,24,Education,No,586,25.52,2.10,0


(25000, 22)

Summary of the Dataset

# Dataset Overview
Dataset have 25000 rows and 22 coloumns
Each row represent one customer 
This dataset contains information about personal loan applicants and their financial profiles, credit behavior, loan details, and loan default risk.


The dataset can be broadly divided into the following categories:

### 1. Applicant Information

This includes demographic and personal details such as age, gender, marital status, state, and city tier.

### 2. Employment and Banking Information

This section contains information related to the applicant's employment type, monthly income, and banking history.

### 3. Financial and Credit Information

These variables provide information about existing loans, monthly EMI obligations, credit utilization, credit inquiries, late payments, and CIBIL score.

### 4. Loan Information

This includes details about the requested loan, such as loan amount, tenure, purpose, collateral, and interest rate.

### 5. Default Risk Information

The dataset also includes variables related to the applicant's estimated default risk and the final default outcome.

## Overall Objective

The dataset can be used to analyze the relationship between applicant characteristics, financial behavior, credit history, and the risk of loan default.


# Column wise Description

#1.> applicant_id-> this column represents the unique customer id of each customer

#2.> age-> contains age of the customer

#3.> gender-> contains gender of the person

#4.> marital_status-> contains marital status of the customer whether he or she is married/single/divorced

#5.> city_tier-> shows the category of city where the customer lives

#6.> state-> shows in which state of the country the customer lives

#7.> employment_type-> shows the employment type of the customer whether they are salaried, govt employee, private employee, business owner or self employed

#8.> bank_account_vintage_years-> shows from how many years the bank account of a particular customer is opened

#9.> monthly_income_inr-> shows the monthly income of the customer in INR

#10.> existing_loans_count-> number of active loans per customer

#11.> existing_emi_monthly_inr-> the monthly EMI of the active loans in INR per customer

#12.> credit_utilization_ratio_pct-> shows how much percentage of available credit is being used by an individual customer

#13.> num_credit_inquiries_last_6m-> number of times the customer's credit history was checked for loans or credit related applications in the last 6 months

#14.> late_payments_last_12m-> number of times the customer made a late payment in the last 12 months

#15.> loan_amount_requested_inr-> how much loan is requested to the bank by different customers

#16.> loan_tenure_months-> shows the duration of the loan repayment period in months

#17.> loan_purpose-> the reason why the customer requested the loan from the bank

#18.> collateral_provided-> whether the customer provided collateral as security to the bank or not

#19.> cibil_score-> the CIBIL score of each customer

#20.> interest_rate_offered_pct-> interest rate offered by the bank to the customer for the loan repayment

#21.> default_risk_score-> a numerical score that represents the estimated risk of a customer defaulting on the loan

#22.> default_flag-> shows whether the customer defaulted on the loan or not
"0"-> means no default
"1"-> means customer is defaulted by the bank

In [3]:
display(Loan_df.head(5))
display(Loan_df.tail(5))
display(Loan_df.sample(5))
display(Loan_df.info())
display(Loan_df.describe())
display(Loan_df.isnull().sum())
display(Loan_df.duplicated().sum())

,applicant_id,age,gender,marital_status,city_tier,state,employment_type,bank_account_vintage_years,monthly_income_inr,existing_loans_count,...,num_credit_inquiries_last_6m,late_payments_last_12m,loan_amount_requested_inr,loan_tenure_months,loan_purpose,collateral_provided,cibil_score,interest_rate_offered_pct,default_risk_score,default_flag
0,APP106868,22,Male,Single,Tier 3,Haryana,Salaried - PSU,0.5,12000,1,...,2.0,0,149000,48,Personal Expense,No,649,24.84,1.67,0
1,APP124016,35,Female,Divorced,Tier 2,Gujarat,Salaried - Private,0.5,32500,3,...,0.0,1,156000,60,Personal Expense,No,538,25.25,4.47,0
2,APP109668,41,Male,Married,Tier 2,Uttar Pradesh,Salaried - Government,2.8,52500,0,...,0.0,0,70000,36,Wedding,No,636,20.46,0.19,0
3,APP113640,29,Female,Married,Tier 2,Kerala,Salaried - PSU,0.7,15500,1,...,0.0,1,108000,48,Education,No,721,15.27,0.26,0
4,APP114018,31,Male,Married,Tier 1,Punjab,Self-Employed Professional,4.7,269600,2,...,0.0,2,466000,24,Education,No,586,25.52,2.10,0


,applicant_id,age,gender,marital_status,city_tier,state,employment_type,bank_account_vintage_years,monthly_income_inr,existing_loans_count,...,num_credit_inquiries_last_6m,late_payments_last_12m,loan_amount_requested_inr,loan_tenure_months,loan_purpose,collateral_provided,cibil_score,interest_rate_offered_pct,default_risk_score,default_flag
24995,APP121575,29,Male,Single,Tier 1,Andhra Pradesh,Salaried - Private,0.7,85900,1,...,1.0,1,113000,24,Personal Expense,No,594,25.24,1.00,0
24996,APP105390,49,Female,Single,Tier 1,Punjab,Self-Employed Professional,0.9,142600,1,...,2.0,2,35000,24,Debt Consolidation,No,638,19.78,0.56,0
24997,APP100860,26,Other,Divorced,Tier 1,Haryana,Salaried - Private,3.0,76600,1,...,2.0,0,67000,36,Wedding,No,562,19.11,4.03,0
24998,APP115795,44,Female,Married,Tier 1,Gujarat,Self-Employed Professional,7.3,68200,1,...,2.0,0,296000,84,Medical Emergency,No,665,16.23,0.45,0
24999,APP123654,34,Female,Married,Tier 3,Delhi NCR,Business Owner,21.2,72900,1,...,2.0,1,172000,36,Wedding,No,710,13.54,0.35,0


,applicant_id,age,gender,marital_status,city_tier,state,employment_type,bank_account_vintage_years,monthly_income_inr,existing_loans_count,...,num_credit_inquiries_last_6m,late_payments_last_12m,loan_amount_requested_inr,loan_tenure_months,loan_purpose,collateral_provided,cibil_score,interest_rate_offered_pct,default_risk_score,default_flag
1312,APP116454,32,Male,Single,Tier 2,Madhya Pradesh,Salaried - PSU,1.0,23500,1,...,2.0,2,109000,72,Home Renovation,No,617,20.47,2.97,0
9315,APP108486,34,Female,Divorced,Tier 2,Haryana,Salaried - Private,0.5,64500,0,...,3.0,2,166000,72,Debt Consolidation,No,564,24.94,3.06,0
7438,APP120771,44,Male,Married,Tier 1,Tamil Nadu,Salaried - Private,1.1,52200,0,...,2.0,2,96000,12,Debt Consolidation,No,595,19.47,2.84,0
5275,APP105589,26,Male,Single,Tier 3,Karnataka,Salaried - Private,1.4,13500,2,...,2.0,1,150000,84,Debt Consolidation,Yes,584,22.04,5.67,0
10442,APP117920,45,Male,Married,Tier 2,Haryana,Salaried - Private,0.5,74500,1,...,1.0,5,126000,24,Medical Emergency,Yes,559,19.62,2.32,0


<class 'pandas.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 22 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   applicant_id                  25000 non-null  str    
 1   age                           25000 non-null  int64  
 2   gender                        25000 non-null  str    
 3   marital_status                25000 non-null  str    
 4   city_tier                     25000 non-null  str    
 5   state                         25000 non-null  str    
 6   employment_type               25000 non-null  str    
 7   bank_account_vintage_years    24625 non-null  float64
 8   monthly_income_inr            25000 non-null  int64  
 9   existing_loans_count          25000 non-null  int64  
 10  existing_emi_monthly_inr      24500 non-null  float64
 11  credit_utilization_ratio_pct  25000 non-null  float64
 12  num_credit_inquiries_last_6m  24750 non-null  float64
 13  late_payment

None

,age,bank_account_vintage_years,monthly_income_inr,existing_loans_count,existing_emi_monthly_inr,credit_utilization_ratio_pct,num_credit_inquiries_last_6m,late_payments_last_12m,loan_amount_requested_inr,loan_tenure_months,cibil_score,interest_rate_offered_pct,default_risk_score,default_flag
count,25000.000000,24625.000000,25000.000000,25000.000000,24500.000000,25000.000000,24750.000000,25000.000000,2.500000e+04,25000.000000,25000.000000,25000.000000,25000.000000,25000.000000
mean,35.303520,3.999647,54832.428000,1.103720,11705.636735,41.590360,1.298869,1.446680,1.584572e+05,42.401760,615.290640,20.714922,6.271094,0.061400
std,8.561309,3.900615,38347.488163,1.051421,10297.365558,19.804901,1.136458,1.271071,1.042315e+05,19.223168,63.057171,3.413218,11.938843,0.240067
min,21.000000,0.500000,12000.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.500000e+04,12.000000,360.000000,10.570000,0.000000,0.000000
25%,29.000000,1.100000,29500.000000,0.000000,5100.000000,27.800000,0.000000,0.000000,8.800000e+04,24.000000,573.750000,18.810000,0.530000,0.000000
50%,35.000000,2.800000,45100.000000,1.000000,9000.000000,41.300000,1.000000,1.000000,1.330000e+05,36.000000,617.000000,21.050000,1.770000,0.000000
75%,41.000000,5.500000,68700.000000,2.000000,15100.000000,55.000000,2.000000,2.000000,1.990000e+05,60.000000,659.000000,23.510000,5.852500,0.000000
max,65.000000,25.000000,727000.000000,6.000000,202300.000000,100.000000,9.000000,8.000000,1.519000e+06,84.000000,866.000000,26.000000,99.080000,1.000000


applicant_id                      0
age                               0
gender                            0
marital_status                    0
city_tier                         0
state                             0
employment_type                   0
bank_account_vintage_years      375
monthly_income_inr                0
existing_loans_count              0
existing_emi_monthly_inr        500
credit_utilization_ratio_pct      0
num_credit_inquiries_last_6m    250
late_payments_last_12m            0
loan_amount_requested_inr         0
loan_tenure_months                0
loan_purpose                      0
collateral_provided               0
cibil_score                       0
interest_rate_offered_pct         0
default_risk_score                0
default_flag                      0
dtype: int64

np.int64(0)

#Data Assesment

#Dirty Data
 there are some runs in which existing loan count=0 but there are s0me value in existing emi column which is not possible
#there are 7920 rows where existing loan counts=0 but there is a existing montly emi -> validity issue


#there are 7 rows where customer age is less than account age -> validity  issue


there are 375 missing values in bank account vintage years coloumns -> completeness issue

there are 500 missing values in existing emi_montly_inr counts coloumn -> completeness issue

there are 250 missing values in num_credits_inequirs_last 6m columns -> completeness issue











In [4]:
display(Loan_df["applicant_id"].duplicated().sum())
#there are no duplicated values in applicant_id column (it should be unique)
display(Loan_df.duplicated().sum())
#"0" that means no 2 rows are repeated it is a good news


np.int64(0)

np.int64(0)

In [5]:
Loan_df[Loan_df["monthly_income_inr"]<Loan_df["existing_emi_monthly_inr"]]
#There are no rows where existing monthly EMI is greater than monthly income


,applicant_id,age,gender,marital_status,city_tier,state,employment_type,bank_account_vintage_years,monthly_income_inr,existing_loans_count,...,num_credit_inquiries_last_6m,late_payments_last_12m,loan_amount_requested_inr,loan_tenure_months,loan_purpose,collateral_provided,cibil_score,interest_rate_offered_pct,default_risk_score,default_flag


In [6]:
Loan_df[Loan_df["loan_amount_requested_inr"]>Loan_df["monthly_income_inr"]].shape

#in a data of 25000 customers there are "22324" customers in which there amount requested is greater than thier montly income



(22324, 22)

In [7]:
display(Loan_df[Loan_df["age"]<=0].shape)
gender=["Male","Female","Other"]
display(Loan_df[~Loan_df["gender"].isin(gender)].shape)
display(Loan_df[Loan_df['bank_account_vintage_years']<0].shape)
display(Loan_df[Loan_df["monthly_income_inr"]<=0].shape)
display(Loan_df[Loan_df["existing_loans_count"]<0].shape)
display(Loan_df[Loan_df["existing_emi_monthly_inr"]<0].shape)
display(Loan_df[Loan_df["num_credit_inquiries_last_6m"]<0].shape)
display(Loan_df[Loan_df["late_payments_last_12m"]<0].shape)
display(Loan_df[Loan_df["loan_amount_requested_inr"]<0].shape)
display(Loan_df[Loan_df["loan_tenure_months"]<=0].shape)

display(Loan_df[
    (Loan_df["cibil_score"] < 300) |
    (Loan_df["cibil_score"] > 900)
].shape)

display(Loan_df[Loan_df["interest_rate_offered_pct"]<=0].shape)
Loan_df[
    Loan_df["num_credit_inquiries_last_6m"].notna() &
    (Loan_df["num_credit_inquiries_last_6m"] % 1 != 0)
]


df=Loan_df[Loan_df["existing_loans_count"]==0]
display(df[df["existing_emi_monthly_inr"]>0].shape)
#there are 7920 rows where existing loan counts=0 but there is a existing montly emi
Loan_df[Loan_df["age"]<Loan_df["bank_account_vintage_years"]].shape
#there are 7 rows where customer age ia less than account age

Loan_df[
    (Loan_df["age"] < 18) &
    (Loan_df["existing_emi_monthly_inr"] > 0)
].shape



(0, 22)

(0, 22)

(0, 22)

(0, 22)

(0, 22)

(0, 22)

(0, 22)

(0, 22)

(0, 22)

(0, 22)

(0, 22)

(0, 22)

(7920, 22)

(0, 22)